torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
+    std::vector<at::Tensor>& inputTensors,
+    std::vector<int>& outputSplitSizes,
+    std::vector<int>& inputSplitSizes,
+    const AllToAllOptions& opts) {
+  checkSingleTensor(outputTensors);
+  checkSingleTensor(inputTensors);
+
+  if(outputSplitSizes.size() == 0 && inputSplitSizes.size() == 0) {
+    // We can use alltoall
+    if ((outputTensors[0].numel() != inputTensors[0].numel()) ||
+        (outputTensors[0].type() != inputTensors[0].type())) {
+      throw std::runtime_error("Tensors are not equal in size or data type");
+    }
+    checkSingleTensorHelper(inputTensors[0]);
+    checkSingleTensorHelper(outputTensors[0]);
+    if(outputTensors[0].size(0) % size_ != 0) {
+      throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+    }
+    
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoall(
+              srcdata.data_ptr(),
+              srcdata.numel()/size_,
+              mpiDatatype.at(srcdata.scalar_type()),
+              dstdata.data_ptr(),
+              dstdata.numel()/size_,
+              mpiDatatype.at(dstdata.scalar_type()),
+              pgComm_));
+      };
+    auto entry = std::unique_ptr<WorkEntry>(
+        new WorkEntry(&inputTensors, &outputTensors, std::move(runFunc)));
+    return enqueue(std::move(entry));
+  } else {
+    // Need alltoallv
+    auto checkSplitSizes = [this](std::vector<int> split_sizes, at::Tensor &tensor) {
+      if(split_sizes.size() == 0) {
+        if(tensor.size(0) % size_ != 0) {
+          throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+        }
+      } else {
+        if(split_sizes.size() != size_) {
+          throw std::runtime_error("Number of tensor splits not equal to group size");
+        }
+        int sum = std::accumulate(split_sizes.begin(), split_sizes.end(), 0);
+        if(sum != tensor.size(0)) {
+          throw std::runtime_error("Split sizes doesn't match total dim 0 size");
+        }
+      }
+    };
+    checkSplitSizes(inputSplitSizes, inputTensors[0]);
+    checkSplitSizes(outputSplitSizes, outputTensors[0]);
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this, inputSplitSizes, outputSplitSizes](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        std::vector<int> send_lengths(size_);
+        std::vector<int> recv_lengths(size_);
+        std::vector<int> send_offsets(size_);
+        std::vector<int> recv_offsets(size_);
+        auto computeLengthsAndOffsets = [this](std::vector<int> split_sizes, at::Tensor &tensor, std::vector<int> &lengths, std::vector<int> &offsets) {
```

Comment:
Is it idiomatic to use lambda functions this way? You can just hoist this as a utility method in this class.

Reply:
Ideally it should be an utility method as functionality is useful for other APIs as well if we decide to extend them with `XX_base` versions and add support for uneven split. But for now it is required only in alltoall implementation, so I localized it as lambda function.

Reply:
My concern is that it pollutes the code somewhat arguably making it less readable, how about a static helper function just outside ProcessGroupMPI::alltoall_base so we don't need to do that move later?

Reply:
Cleaner is to make these utility methods. Will move them near `checkSingleTensor`.